From b642471f2ed2f1a2b1f25ef72e941f7eef9ec8d0 Mon Sep 17 00:00:00 2001
From: Subroto Sanyal <ssanyal@datameer.com>
Date: Fri, 3 Jun 2016 16:50:00 -0700
Subject: [PATCH 206/517] [SPARK-15754][YARN] Not letting the credentials
 containing hdfs delegation tokens to be added in
 current user credential.

## What changes were proposed in this pull request?
The credentials are not added to the credentials of UserGroupInformation.getCurrentUser(). Further if the client has possibility to login using keytab then the updateDelegationToken thread is not started on client.

## How was this patch tested?
ran dev/run-tests

Author: Subroto Sanyal <ssanyal@datameer.com>

Closes #13499 from subrotosanyal/SPARK-15754-save-ugi-from-changing.

(cherry picked from commit 61d729abdaf9bdd54aea74a643828d570b036d87)
---
 .../org/apache/spark/deploy/yarn/Client.scala      |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala b/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala
index 719f1e2..e87c51e 100644
--- a/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala
+++ b/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala
@@ -883,7 +883,6 @@ private[spark] class Client(
       CryptoConf.initSparkShuffleCredentials(sparkConf, credentials)
     }
     setupSecurityToken(amContainer)
-    UserGroupInformation.getCurrentUser().addCredentials(credentials)
 
     amContainer
   }
@@ -911,7 +910,8 @@ private[spark] class Client(
       sparkConf.set("spark.yarn.keytab", keytabFileName)
       sparkConf.set("spark.yarn.principal", principal)
     }
-    credentials = UserGroupInformation.getCurrentUser.getCredentials
+    // Defensive copy of the credentials
+    credentials = new Credentials(UserGroupInformation.getCurrentUser.getCredentials)
   }
 
   /**
-- 
1.7.9.5

